Events

aapeliv opened the pull request:
This is now ready for review, hoping to get it merged in.

After that, the following (big chunks) will be in later PRs:

* Proper timezone handling
* Notifications
* Corresponding `ListEvents` in communities, groups
* Search functionality
* Automatically repeating events (currently need to schedule each instance manually) — not so urgent

It was too big to do in one PR.

aapeliv committed:
Start on events

aapeliv committed:
WIP

aapeliv committed:
Get a rough implementation working

aapeliv committed:
Clean up

aapeliv committed:
Fix basic errors

aapeliv committed:
Fix migrations

aapeliv committed:
Fix migration

aapeliv committed:
WIP

aapeliv committed:
WIP

aapeliv committed:
Work on events

aapeliv committed:
Tweak

aapeliv committed:
Fix migrations and other tweaks

aapeliv committed:
Work on tests

aapeliv committed:
Work on event tests

aapeliv committed:
Add more tests and fix related bugs

aapeliv committed:
More tests

aapeliv committed:
Clean up

aapeliv committed:
Fix pagination bug.

The page token accuracy was millis, but postgres accuracy is more, so this was causing the pagination to be flaky depending on rounding up or down pg ts.

aapeliv committed:
Write more tests

aapeliv committed:
More tests

aapeliv committed:
Enforce online/offline dichotomy properly

aapeliv committed:
Fix up update test

aapeliv committed:
Add a test to update all future events

aapeliv committed:
Test changing to offline

aapeliv committed:
Add tests

kalvdans commented on app/backend/src/couchers/errors.py:
```diff
@@ -27,6 +27,17 @@
+EVENT_CANT_UPDATE_OLD_EVENT = "You can't update an event that's more than 1 day old."
+EVENT_EDIT_PERMISSION_DENIED = "You're not allowed to edit that event."
+EVENT_ENDS_BEFORE_STARTS = "The event must end after it starts"
+EVENT_IN_PAST = "The event must be in the future"
```
Some of the added error strings does not end with a dot.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1029,30 +1037,116 @@ class ClusterEventAssociation(Base):
+    __table_args__ = (
+        # Only one of owner_user and owner_cluster should be set
+        CheckConstraint(
+            "(owner_user_id IS NULL AND owner_cluster_id IS NOT NULL) OR (owner_user_id IS NOT NULL AND owner_cluster_id IS NULL)",
```
Can be written shorter `((owner_user_id ISNULL) != (owner_cluster_id ISNULL))` (untested)

aapeliv replied:
I think as it's written now, it's clearer.

(We have the same type of constraints in several places, I think.)

Too bad pg doesn't have an XOR operator

kalvdans replied:
`!=` is xor when applied to boolean values.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1029,30 +1037,116 @@ class ClusterEventAssociation(Base):
+    id = Column(BigInteger, communities_seq, primary_key=True, server_default=communities_seq.next_value())
+    event_id = Column(ForeignKey("events.id"), nullable=False, index=True)
+
+    creator_user_id = Column(ForeignKey("users.id"), nullable=False, index=True)
```
How is creator_user_id different from Event.creator_user_id ? Please write a comment.

kalvdans commented on app/backend/src/couchers/models.py:
```diff
@@ -1029,30 +1037,116 @@ class ClusterEventAssociation(Base):
+    # videoconferencing link, etc, must be specified if no geom, otherwise optional
+    link = Column(String, nullable=True)
+
+    timezone = "Etc/UTC"
```
capitalize constants to not confuse them with a Column name.

aapeliv replied:
Idea is to turn this into a column in next pr without too much hassle

kalvdans replied:
there will be hassle anyway but resolving conversation.

kalvdans commented on app/pb/events.proto:
```diff
@@ -0,0 +1,304 @@
+    // Lists all subscribed, attending
+  }
+
+  // TODO: DownloadCalendarInvite
```
Client-side

aapeliv replied:
My idea was to generate the file (an icalendar, `.ics`) on the backend and just send that to the frontends. It's a standard format for describing events, and I think much better to do that in the backend instead of having to replicate it to each frontend implementation. There will probably also be some details that are just easier to do in the backend (e.g. tz support/conversions).

kalvdans replied:
Or we can provide a javascript function to generate an ics file from GetEvent response, that all frontends can use.

aapeliv replied:
Not all frontends will be written in javascript.

I am also hoping to have better integration with calendar software eventually. It's possible to add a URL that some software will update the event from, so we could potentially push new scheduled occurences directly to the user's calendar. This is not something we can do client side.

Also, client side icalendar handling libraries seem to be pretty crappy compared to https://pypi.org/project/icalendar/

kalvdans replied:
I disagree but both ways have their merits, so resolving conversation :)

kalvdans commented on app/pb/events.proto:
```diff
@@ -0,0 +1,304 @@
+
+message Event {
+  int64 event_id = 1; // = occurence_id
+  // it's the next event occurence
+  bool is_next = 2;
```
Please define "next". Is it only in my current polygon and does already started occurrences count?

aapeliv replied:
Thanks, great catch

kalvdans commented on app/pb/events.proto:
```diff
@@ -0,0 +1,304 @@
+  int64 event_id = 1; // = occurence_id
+  // it's the next event occurence
+  bool is_next = 2;
+  // already happened
+  bool is_past = 3;
+  // in the future
+  bool is_future = 4;
```
Redundant, client can calculate these from start_date and end_date.

kalvdans commented on app/backend/src/tests/test_events.py:
```diff
@@ -0,0 +1,1726 @@
+        assert res.offline_information.lat == 0.1
+        assert res.offline_information.lng == 0.2
+        assert res.offline_information.address == "Near Null Island"
+        assert time_before < to_aware_datetime(res.created) < now()
```
`<=`

kalvdans commented on app/backend/src/tests/test_events.py:
```diff
@@ -0,0 +1,1726 @@
+        assert res.offline_information.lng == 0.2
+        assert res.offline_information.address == "Near Null Island"
+        assert time_before < to_aware_datetime(res.created) < now()
+        assert time_before < to_aware_datetime(res.last_edited) < now()
```
`<=` to not fail on a fast computer with 100 Hz clock accuracy.

kalvdans commented on app/backend/src/tests/test_events.py:
```diff
@@ -0,0 +1,1726 @@
+            )
+        )
+
+        assert res.is_next
```
Can two simultaneous events both be "next"?

aapeliv replied:
No. I've thought about it, and I think occurences of the same event should not overlap.

kalvdans commented on app/pb/events.proto:
```diff
@@ -0,0 +1,304 @@
+  google.protobuf.Timestamp start_time = 16;
+  google.protobuf.Timestamp end_time = 17;
+
+  // the tzdata timezone identified
+  string timezone = 18;
```
identified? I think this is the timezone decided when creating the event.

aapeliv replied:
It's a typo, should be "identifier", thanks for spotting

kalvdans reviewed (changes requested):
Looks solid in general, some minor comments only.

aapeliv committed:
Fix some review comments

aapeliv committed:
More review comments

aapeliv committed:
Fix migrations

aapeliv committed:
Add comment about creator user

aapeliv commented:
@kalvdans: to clarify, there are events (`Event`) and occurences (`EventOccurence`) in the backend representation. An event is something like "Weekly NYC Couchers meetup", and an occurence is an instance of that, e.g. the time this Friday that the NYC meetup goes to meet up. The purpose for having these two parts is to that users can subscribe to one "stream" of events (through `EventSubscription`) and be notified (in future PR) of new occurences of an event. So if I'm in the NYC Couchers meetup, I want to get an email every week telling me where we're meeting up and at what time, who's going (`EventOccurenceAttendance`, separately from who is subscribed), etc.

Please let me know if this is not clear in comments (I guess it's not: sorry). Where should I document this? I can add it in something like `docs/events.md` or in models? protos (but this is backend implementation details)? I'm not sure.

The frontends/API layer just has one "event" object, and the way to get info about other related events is through `ListEventOccurences`. That's why there's `is_next`.

Regarding `is_next`: I think the right way is to add an exclusion constraint to make sure there is only one event in a given stream at a time, no overlaps.

Thoughts?

kalvdans commented:
> Thoughts?

I'm all for having a lot of comments in the .proto files, like your story with "Weekly NYC Couchers meetup". So the documentation for is_next is very much good now but can be clarified that it is always true for non-recurring events.

Implementation details, like having two tables (one for event series and one for event occurrances) should go into `models.py` in my opinion.

aapeliv committed:
Fix migrations

aapeliv committed:
Switch to using a tstzrange to represent event time and add constraint that occurences within the same event cannot overlap

aapeliv committed:
Add nice error message for scheduling overlapping event

aapeliv committed:
Add error message for update

aapeliv committed:
Simplify all check constraints

Breaks migrations, will fix soon

aapeliv committed:
Fix

aapeliv committed:
Add migrations for constraints

aapeliv commented:
I've changed this to use a tstzrange (timestamp with timezone range) to represent the time during which the event takes place, and then used that to make sure there is only one occurence at a time. This fixes up the is_next bit.

I also cleaned up all existing constraints.

aapeliv committed:
Clean up

aapeliv commented:
Ready for re-review @kalvdans

aapeliv committed:
Fix bounds for event duration

kalvdans commented on app/pb/events.proto:
```diff
@@ -0,0 +1,301 @@
+  }
+
+  rpc ScheduleEvent(ScheduleEventReq) returns (Event) {
+    // Schedules a new occurence of an existing event
```
occurrence is spelled with two r (here and more places)

aapeliv replied:
Wow... Yeah, I completely didn't notice. Let's hope find & replace suffices!

aapeliv committed:
Fix spelling of occurrence

kalvdans commented on app/backend/src/couchers/migrations/versions/c035cdcabd40_add_new_events_tables.py:
```diff
@@ -75,7 +75,7 @@ def upgrade():
     )
     op.create_table(
-        "event_occurences",
+        "event_occurrences",
```
Don't touch old migrations! (your own words :) )

aapeliv replied:
This is a new migration in this PR so it's fine. But yes, shouldn't touch migrations already on `develop` (or a bit more leniently, those that have been applied in prod)

kalvdans replied:
My mistake looking at the last commit only 😊

kalvdans reviewed (changes requested):
Aww, the spelling checks made it hard to see the actual changes. Anyway the is_next documentation is clearer now, thanks. I think it looks ok besides some new issues introduced by the spelling corrections.

We might live with wrongly spelled tables and fix that separately later.

aapeliv commented:
> I think it looks ok besides some new issues introduced by the spelling corrections.

I think it's all fine. This is the first PR to do events (there were some existing tables but no data/code that used them).

So there shouldn't be any remnants of wrong spelling. (There is one occurrence of "occurence" in the migrations to do with `slugify`, and I didn't change that, even though it just seems to be a comment).
